README.md
# alt-text ![Build Status](https://github.com/double-great/alt-text/workflows/Test/badge.svg)

Check for unhelpful alt text.

Alt text describes an image. This description provides context for people with low or no vision. The text is a fallback for users with images turned off or when an image fails to load.

The usefulness of alt text can be subjective. Context, detail, length, and relevance contribute to the quality of the description. Determining alt text quality is a manual effort, but there are patterns to avoid.

This library checks for common issues found in alt text and suggests improvements.

Are you writing in markdown? We recommend using [remark-lint-alt-text](https://github.com/double-great/remark-lint-alt-text).

## Install

```
npm install @double-great/alt-text
```

## Usage

If the library detects an issue, it will return a suggestion to help you fix it. If it doesn't detect any issues it will return `undefined`.

```js
import altText from "@double-great/alt-text";

console.log(altText("A child holding a photograph."));
// undefined

console.log(altText("A photo of a dog"));
// Alt text should not contain "photo of" (https://tinyurl.com/ydbwu53g). Alt text should end with punctuation (https://tinyurl.com/y9fcquhy).
```

The library is configured with best practices, but context is always important. You can disable any rule and pass custom configuration options for those that check against specific values.

Example:

```js
console.log(altText("A photo of a dog", { "end-with-punctuation": false }));
// Alt text should not contain "photo of" (https://tinyurl.com/ydbwu53g).
```

Each rule in the suggestions list below has full documentation on configuration options.

## Suggestions

<!-- this section is generated on commit !-->

- [Alt text contains unhelpful words (`contains-unhelpful-word`)](#alt-text-contains-unhelpful-words)
- [Alt text is unhelpful (`is-unhelpful`)](#alt-text-is-unhelpful)
- [Alt text should not end with (`should-not-end-with`)](#alt-text-should-not-end-with)
- [Alt text should not start with (`should-not-start-with`)](#alt-text-should-not-start-with)
- [Avoid emoji (`avoid-emoji`)](#avoid-emoji)
- [Character length (`character-length`)](#character-length)
- [Empty alt text (`empty-alt-text`)](#empty-alt-text)
- [End with punctuation (`end-with-punctuation`)](#end-with-punctuation)
- [Image is decorative (`image-is-decorative`)](#image-is-decorative)
- [Image is link (`image-is-link`)](#image-is-link)
- [Missing alt attribute (`no-alt`)](#missing-alt-attribute)

### Alt text contains unhelpful words

Suggestion: `Alt text should not contain "graphic of, image of, photo of, photo:, photograph of, photographer:, picture of, screen shot of, screenshot of"`

Screen readers announce the presence of an image before reading the alt text. Adding “picture of” or “photo of” is redundant in this context.

- ✅ Dog jumping through a hoop.
- 🚫 Image of a dog jumping through a hoop.

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"contains-unhelpful-word":false})
// adjust rule defaults:
altText("My alt text.", {"contains-unhelpful-word":{"exclude":["graphic of","image of","photo of","photo:","photograph of","photographer:","picture of","screen shot of","screenshot of"]}})
```
<!-- prettier-ignore-end -->

Sources:

- <https://www.w3.org/WAI/tutorials/images/tips/#tips>
- <https://axesslab.com/alt-texts/#dont-say-its-an-image>

### Alt text is unhelpful

Suggestion: `Alt text should not be "*, alt, arrow, artwork, blank, bullet, button, chart, diagram, drawing, empty, graph, graphic, icon, image, logo, more, painting, photo, photograph, placeholder, screen shot, screenshot, spacer, table, temp"`

Usually, there’s no need to include words like “image”, “icon”, or “picture” in the alt text. People who can see will know this already, and screen readers announce the presence of an image.

- ✅ A child holding a photograph.
- 🚫 photograph

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"is-unhelpful":false})
// adjust rule defaults:
altText("My alt text.", {"is-unhelpful":{"exclude":["*","alt","arrow","artwork","blank","bullet","button","chart","diagram","drawing","empty","graph","graphic","icon","image","logo","more","painting","photo","photograph","placeholder","screen shot","screenshot","spacer","table","temp"]}})
```
<!-- prettier-ignore-end -->

Sources:

- <https://www.w3.org/WAI/tutorials/images/tips/#tips>
- <https://axesslab.com/alt-texts/#dont-say-its-an-image>

### Alt text should not end with

Suggestion: `Alt text should not end with ".gif, .jpeg, .jpg, .png, .svg, .webp, graphic, image"`

A file name in alt text does not provide helpful context.

- ✅ A child holding a photograph.
- 🚫 photograph.jpg

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"should-not-end-with":false})
// adjust rule defaults:
altText("My alt text.", {"should-not-end-with":{"exclude":[".gif",".jpeg",".jpg",".png",".svg",".webp","graphic","image"]}})
```
<!-- prettier-ignore-end -->

Sources:

- <https://axesslab.com/alt-texts/>

### Alt text should not start with

Suggestion: `Alt text should not start with "graphic, image, photo, photograph, photographer, picture, screen shot, screenshot, spacer"`

Usually, there’s no need to include words like “image”, “icon”, or “picture” in the alt text. People who can see will know this already, and screen readers announce the presence of an image.

- ✅ A child holding a photograph.
- 🚫 Image of a child.

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"should-not-start-with":false})
// adjust rule defaults:
altText("My alt text.", {"should-not-start-with":{"exclude":["graphic","image","photo","photograph","photographer","picture","screen shot","screenshot","spacer"]}})
```
<!-- prettier-ignore-end -->

Sources:

- <https://www.w3.org/WAI/tutorials/images/tips/#tips>
- <https://axesslab.com/alt-texts/#dont-say-its-an-image>

### Avoid emoji

Suggestion: `Replace emoji in alt text with descriptive text`

Emoji have their own text descriptions. These descriptions can vary between operating systems and software. The spoken description of the emoji may not match your visual intention.

- ✅ `<img src="cat.jpg" alt="An orange cat.">`
- 🚫 `<img src="cat.png" alt="An orange 🐈."/>`

Hear an example: <https://doublegreat.dev/listen/emoji/>

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"avoid-emoji":false})
```
<!-- prettier-ignore-end -->

Sources:

- <https://www.youtube.com/watch?v=uIbPcZq6izk>
- <https://readabilityguidelines.co.uk/images/emojis/>

### Character length

Suggestion: `Alt text length should be less than 125 characters`

Alt text should be less than 125 characters in length. The JAWS screen reader reads alt text in 125 character chunks. Any information about the image, such as copyright information, image source or extra information should be placed in the caption text below the image.

- ✅ George Washington and Lafayette on horseback talking to soldiers in snow at Valley Forge.
- 🚫 Caption: Painting “Washington and Lafayette at Valley Forge” by John Ward Dunsmore from 1907. Image courtesy of the Library of Congress.

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"character-length":false})
// adjust rule defaults:
altText("My alt text.", {"character-length":{"length":125}})
```
<!-- prettier-ignore-end -->

Sources:

- <https://accessibility.psu.edu/images/imageshtml/#alt>
- <https://terrillthompson.com/tests/altlength.html>

### Empty alt text

Suggestion: `Alt text should not be a single space`

If you use null (empty) alt text (`alt=""`) to hide decorative images, make sure that there is no space character in between the quotes. **If a space character is present, the image may not be effectively hidden from assistive technologies.** For instance, some screen readers will still announce the presence of an image if a space character is put between the quotes.

- ✅ `<img src="photo.png" alt="">`
- 🚫 `<img src="photo.png" alt=" ">`

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"empty-alt-text":false})
```
<!-- prettier-ignore-end -->

Sources:

- <https://www.w3.org/WAI/tutorials/images/tips/#tips>

### End with punctuation

Suggestion: `Alt text should end with punctuation`

End the alt text with a period, exclamation point, or question mark. This will make screen readers pause a bit after the last word in the alt text, which creates a more pleasant reading experience for the user.

- ✅ A child holding a photograph.
- 🚫 A child holding a photograph

Hear an example: <https://doublegreat.dev/listen/punctuation-in-alt-text/>

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"end-with-punctuation":false})
```
<!-- prettier-ignore-end -->

Sources:

- <https://axesslab.com/alt-texts/#end-with-a-period>

### Image is decorative

Suggestion: `Empty alt text should only be used for decorative images`

Provide "null" `alt` attributes (using `alt=""`) for images which do not provide information or do not require alternative text because the image is described in the page content. Some developers will mistakenly leave off the alt attribute altogether on images which they deem do not need alternatives. This is not helpful to assistive technology users because the assistive technology, such as screen reader, will often read the source attribute (i.e., file name) as the alternative text. To tell assistive technology to ignore an image, use a "blank alternative text" attribute: `alt=""`.

- ✅ `<img src="decorative-photo.jpg" alt="">`
- 🚫 `<img src="quarterly-earnings-chart.png" alt=""/>`

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"image-is-decorative":false})
```
<!-- prettier-ignore-end -->

Sources:

- <https://dequeuniversity.com/rules/axe/3.0/image-alt>

### Image is link

Suggestion: `Images inside a link tag require alt text that describes the purpose of the link`

Images inside a link tag require alt text that describes the purpose of the link.

- ✅ `<a href="https://github.com/double-great"><img src="logo.png" alt="double great on GitHub"></a>`
- 🚫 `<a href="https://github.com/double-great"><img src="logo.png" alt="double great logo"></a>`

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"image-is-link":false})
```
<!-- prettier-ignore-end -->

Sources:

- <https://axesslab.com/alt-texts/#images-in-links>

### Missing alt attribute

Suggestion: `Missing "alt" attribute`

All images must have alternate text to convey their purpose and meaning to screen reader users.

- ✅ `<img src="photograph.jpg" alt="A child holding a photograph.">`
- 🚫 `<img src="photograph.jpg">`

Configuration:

<!-- prettier-ignore-start -->
```js
// disable the rule:
altText("My alt text.", {"no-alt":false})
```
<!-- prettier-ignore-end -->

Sources:

- <https://dequeuniversity.com/rules/axe/3.4/image-alt>


CHANGELOG.md
# Changelog

## 3.1.0

- Remove prettier as a dependency for generating documention.
- Switch to use doublegreat.dev for documentation.

## 3.0.0

- Add configuration options to disable any rule and change the defaults for some rules that check against specific values.

## 2.0.0

- Switch to TypeScript.
- Replace the deprecated git.io with tinyurl.

## 1.0.0

- Switch to ES module.

## 0.6.0

- Add "screen shot" and "screenshot" to exact match. [#63](https://github.com/double-great/alt-text/pull/63)
- Fix stray `console.log`. [#61](https://github.com/double-great/alt-text/pull/61)

## 0.5.0

- Add "Hear an example" section to documentation and include links under "Avoid emoji" and "End with punctuation" suggestions. [#58](https://github.com/double-great/alt-text/pull/58)
- Update dependencies.

## 0.4.0

- Replace "End in period" suggestion with "End in punctuation". [#49](https://github.com/double-great/alt-text/pull/49)

## 0.3.1

- Register "Avoid emoji" suggestion with main script. [#48](https://github.com/double-great/alt-text/pull/48)

## 0.3.0

- Add "Avoid emoji" suggestion. [#46](https://github.com/double-great/alt-text/pull/46)
- Rename all "warnings" as suggestions. [#45](https://github.com/double-great/alt-text/pull/45)


clues.js
export default {
  notOnlySpace: {
    heading: "Empty alt text",
    docs: "https://tinyurl.com/y2o7uctu",
    suggestion: () => `Alt text should not be a single space`,
    rationale:
      'If you use null (empty) alt text (`alt=""`) to hide decorative images, make sure that there is no space character in between the quotes. **If a space character is present, the image may not be effectively hidden from assistive technologies.** For instance, some screen readers will still announce the presence of an image if a space character is put between the quotes.',
    source: ["https://www.w3.org/WAI/tutorials/images/tips/#tips"],
    ok: '`<img src="photo.png" alt="">`',
    notOk: '`<img src="photo.png" alt=" ">`',
  },
  charLength: {
    heading: "Character length",
    docs: "https://tinyurl.com/y2f7rhao",
    suggestion: (length) =>
      `Alt text length should be less than 125 characters${
        length ? `, it is currently ${length} characters` : ""
      }`,
    rationale:
      "Alt text should be less than 125 characters in length. The JAWS screen reader reads alt text in 125 character chunks. Any information about the image, such as copyright information, image source or extra information should be placed in the caption text below the image.",
    source: [
      "https://accessibility.psu.edu/images/imageshtml/#alt",
      "https://terrillthompson.com/tests/altlength.html",
    ],
    ok: "George Washington and Lafayette on horseback talking to soldiers in snow at Valley Forge.",
    notOk:
      "Caption: Painting “Washington and Lafayette at Valley Forge” by John Ward Dunsmore from 1907. Image courtesy of the Library of Congress.",
  },
  imageLink: {
    heading: "Image is link",
    docs: "https://tinyurl.com/yxhq2k5w",
    suggestion: () =>
      `Images inside a link tag require alt text that describes the purpose of the link`,
    rationale:
      "Images inside a link tag require alt text that describes the purpose of the link.",
    source: ["https://axesslab.com/alt-texts/#images-in-links"],
    ok: '`<a href="https://github.com/double-great"><img src="logo.png" alt="double great on GitHub"></a>`',
    notOk:
      '`<a href="https://github.com/double-great"><img src="logo.png" alt="double great logo"></a>`',
  },
  endPunctuation: {
    heading: "End with punctuation",
    docs: "https://tinyurl.com/y5krn3ny",
    suggestion: () => `Alt text should end with punctuation`,
    rationale:
      "End the alt text with a period, exclamation point, or question mark. This will make screen readers pause a bit after the last word in the alt text, which creates a more pleasant reading experience for the user.",
    listen: "https://doublegreat.dev/listen/punctuation-in-alt-text/",
    source: ["https://axesslab.com/alt-texts/#end-with-a-period"],
    ok: "A child holding a photograph.",
    notOk: "A child holding a photograph",
  },
  noAlt: {
    heading: "Missing alt attribute",
    docs: "https://tinyurl.com/yybc6bsy",
    suggestion: () => `Missing "alt" attribute`,
    rationale:
      "All images must have alternate text to convey their purpose and meaning to screen reader users.",
    source: ["https://dequeuniversity.com/rules/axe/3.4/image-alt"],
    ok: '`<img src="photograph.jpg" alt="A child holding a photograph.">`',
    notOk: '`<img src="photograph.jpg">`',
  },
  contains: {
    heading: "Alt text contains unhelpful words",
    docs: "https://tinyurl.com/y3v3jgux",
    fn: (item, alt) => alt.includes(item),
    suggestion: (value) => `Alt text should not contain "${value}"`,
    rules: [
      "picture of",
      "photo of",
      "photograph of",
      "image of",
      "graphic of",
      "screenshot of",
      "screen shot of",
      "photo:",
      "photographer:",
    ],
    rationale:
      "Screen readers announce the presence of an image before reading the alt text. Adding “picture of” or “photo of” is redundant in this context.",
    source: [
      "https://www.w3.org/WAI/tutorials/images/tips/#tips",
      "https://axesslab.com/alt-texts/#dont-say-its-an-image",
    ],
    ok: "Dog jumping through a hoop.",
    notOk: "Image of a dog jumping through a hoop.",
  },
  exactMatch: {
    heading: "Alt text is unhelpful",
    docs: "https://tinyurl.com/yxwc2hof",
    fn: (item, alt) => item == alt.trim(),
    suggestion: (value) => `Alt text should not be "${value}"`,
    rules: [
      "image",
      "icon",
      "graphic",
      "photo",
      "photograph",
      "placeholder",
      "temp",
      "alt",
      "drawing",
      "painting",
      "artwork",
      "logo",
      "bullet",
      "button",
      "arrow",
      "more",
      "spacer",
      "screenshot",
      "screen shot",
      "blank",
      "empty",
      "chart",
      "table",
      "diagram",
      "graph",
      "*",
    ],
    rationale:
      "Usually, there’s no need to include words like “image”, “icon”, or “picture” in the alt text. People who can see will know this already, and screen readers announce the presence of an image.",
    source: [
      "https://www.w3.org/WAI/tutorials/images/tips/#tips",
      "https://axesslab.com/alt-texts/#dont-say-its-an-image",
    ],
    ok: "A child holding a photograph.",
    notOk: "photograph",
  },
  endWith: {
    heading: "Alt text should not end with",
    docs: "https://tinyurl.com/yy2q8bbb",
    fn: (item, alt) => alt.endsWith(item),
    suggestion: (value) => `Alt text should not end with "${value}"`,
    rules: [
      ".jpg",
      ".jpeg",
      ".gif",
      ".png",
      ".svg",
      ".webp",
      "image",
      "graphic",
    ],
    rationale: "A file name in alt text does not provide helpful context.",
    source: ["https://axesslab.com/alt-texts/"],
    ok: "A child holding a photograph.",
    notOk: "photograph.jpg",
  },
  startWith: {
    heading: "Alt text should not start with",
    docs: "https://tinyurl.com/y5y98ygu",
    fn: (item, alt) => alt.startsWith(item),
    suggestion: (value) => `Alt text should not start with "${value}"`,
    rules: [
      "picture",
      "photo",
      "photograph",
      "photographer",
      "image",
      "graphic",
      "screenshot",
      "screen shot",
      "spacer",
    ],
    rationale:
      "Usually, there’s no need to include words like “image”, “icon”, or “picture” in the alt text. People who can see will know this already, and screen readers announce the presence of an image.",
    source: [
      "https://www.w3.org/WAI/tutorials/images/tips/#tips",
      "https://axesslab.com/alt-texts/#dont-say-its-an-image",
    ],
    ok: "A child holding a photograph.",
    notOk: "Image of a child.",
  },
  decorative: {
    heading: "Image is decorative",
    docs: "https://tinyurl.com/yxnvejgv",
    suggestion: () =>
      `Empty alt text should only be used for decorative images`,
    rationale: `Provide "null" \`alt\` attributes (using \`alt=""\`) for images which do not provide information or do not require alternative text because the image is described in the page content. Some developers will mistakenly leave off the alt attribute altogether on images which they deem do not need alternatives. This is not helpful to assistive technology users because the assistive technology, such as screen reader, will often read the source attribute (i.e., file name) as the alternative text. To tell assistive technology to ignore an image, use a "blank alternative text" attribute: \`alt=""\`.`,
    source: ["https://dequeuniversity.com/rules/axe/3.0/image-alt"],
    ok: '`<img src="decorative-photo.jpg" alt="">`',
    notOk: '`<img src="quarterly-earnings-chart.png" alt=""/>`',
  },
  avoidEmoji: {
    heading: "Avoid emoji",
    docs: "https://tinyurl.com/yylrxrus",
    suggestion: (emoji) =>
      `Replace ${emoji || "emoji"} in alt text with descriptive text`,
    rationale: `Emoji have their own text descriptions. These descriptions can vary between operating systems and software. The spoken description of the emoji may not match your visual intention.`,
    listen: "https://doublegreat.dev/listen/emoji/",
    source: [
      "https://www.youtube.com/watch?v=uIbPcZq6izk",
      "https://readabilityguidelines.co.uk/images/emojis/",
    ],
    ok: '`<img src="cat.jpg" alt="An orange cat.">`',
    notOk: '`<img src="cat.png" alt="An orange 🐈."/>`',
  },
};
